### ISSUE TITLE
read_count in email entity never update

### ISSUE THREAD

**Issue by mallezie** [mautic/mautic#13866]
### Mautic Version

5.0.x series

### Way of installing

I installed with composer using https://github.com/mautic/recommended-project

### PHP version

8.1

### What browsers are you seeing the problem on?

Not relevant

### What happened?

When a mail is send out, the is_read entry in the email entity always stays zero. This number is shown on the mail overview, and in the email details tab, where it stays zero. 
Graphs do show the correct number, since they get the data from email_stats table, which is updated correctly.

The issue is AFAICT introduced in M5.1 from following PR. https://github.com/mautic/mautic/pull/12673

Going through the code path, it starts from hitEmail function in EmailModel. https://github.com/mautic/mautic/blob/5.x/app/bundles/EmailBundle/Model/EmailModel.php

Relevant part line 439-451. 
```php
$this->emailStatModel->saveEntity($stat);

        // Only up counts if associated with both an email and lead
        if ($firstTime && $email && $lead) {
            try {
                $this->getRepository()->incrementRead($email->getId(), $stat->getId(), $email->isVariant());
            } catch (\Exception $exception) {
                error_log($exception);
            }
        }
        if ($email) {
            $this->em->persist($email);
        }
```

This updates the stat, followed by updating the email entity itself (the incrementRead function).
However the increment read function uses following query. 

```php
$subQuery = $this->getEntityManager()->getConnection()->createQueryBuilder()
            ->select('es.email_id')
            ->from(MAUTIC_TABLE_PREFIX.'email_stats', 'es')
            ->where('es.id = :statId')
            ->andWhere('es.is_read = 1');

        $q->update(MAUTIC_TABLE_PREFIX.'emails', 'e')
            ->set('read_count', 'read_count + 1')
            ->where(
                $q->expr()->and(
                    $q->expr()->eq('e.id', ':emailId'),
                    $q->expr()->notIn('e.id', $subQuery->getSQL())
                )
            )
            ->setParameter('emailId', $emailId)
            ->setParameter('statId', $statId);

        if ($isVariant) {
            $q->set('variant_read_count', 'variant_read_count + 1');
        }
```

Where it only updates the read_count for mails not yet read in email_stats. Which was done just before we get into here, so we never actually update it. (This filter because of not counting mails being read multiple times). 

Might be enough to switch the stat saving to after the email entity update, but not sure about that. 

### How can we reproduce this issue?

Step 1: Create a mail. 
Step2: Send out mail.
Step 3: Read the mail and see the email is_read column not updated. 

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

<br /><hr>
Care about this issue? Want to get it resolved sooner? If you are a <a href='https://www.mautic.org/become-a-member-of-mautic'>member of Mautic</a>, you can add some funds to the <a href='https://opencollective.com/mautic/projects/bounties'>Bounties Project</a> so that the person who completes this task can claim those funds once it is merged by a member of the core team! Read the docs <a href='https://contribute.mautic.org/product-team/mautic-bounty-programme'>here.</a>

**Comment by avara** [comment]
Hello,

it seems that it could be related with this [issue](https://github.com/mautic/mautic/issues/13807) I created some days ago where I describes also a possible solution.

Regards.

**Comment by mallezie** [comment]
@avara i'm not sure it is entirely related. AFAICT the issue i'm describing here only affects Mautic since 5.1.0

**Comment by avara** [comment]
@mallezie Ok, if it only affects since 5.1.0 it could be something else. But when I read your issue it sounded too similar.

**Comment by patrykgruszka** [comment]
#12673 is related, but it was introduced in 5.0.0. 

@escopecz I think it's this change: https://github.com/mautic/mautic/commit/eb8e81c561de4c06eba42fd88603305093918983#diff-09b3ad8df265e6fc5d43a0ead1c0d5394e085e56093f39af644b641d52827d82R441

Is it necessary to replace the `$this->em->persist($stat);` with `$this->emailStatModel->saveEntity($stat);`?

**Comment by escopecz** [comment]
@patrykgruszka I'm not sure I follow. You suggest to change the code the way it was changed in my PR already?

**Comment by patrykgruszka** [comment]
@escopecz changing the line `$this->em->persist($stat);` to `$this->emailStatModel->saveEntity($stat);` created an issue because the SQL query that increments the read counter relies on the existence of the stat to avoid concurrency problems.

**Comment by escopecz** [comment]
I still don't get it. `$this->emailStatModel->saveEntity($stat);` calls `$this->em->persist($stat);`.

**Comment by mallezie** [comment]
The problem is that first the stat is updated, with $this->emailStatModel->saveEntity($stat);

After that (when it's updated) it checks to see if it needs to update the mail, which by then it does not update it anymore, since the stat is already updated. See the subquery part in updating the mail entity read_count. https://github.com/mautic/mautic/blob/02f3de87c03ba4254650564d78be795af01a05b1/app/bundles/EmailBundle/Entity/EmailRepository.php#L613

**Comment by patrykgruszka** [comment]
It calls `persist()` and the `flush()`.  The flush was called after incrementation of the read counter before this change.

**Comment by escopecz** [comment]
This proves the point that unit tests are useless. Let's not write them anymore especially with dealing with DB queries. Let's write functional tests instead, please. We use the code before https://github.com/mautic/mautic/pull/12673 and it works well. Anyone see any solution?

**Comment by patrykgruszka** [comment]
Indeed. Placing the saving of the stat after the read incrementation should resolve this issue, although I have not tested it yet.

**Comment by avara** [comment]
So, if it's finally as @patrykgruszka said an issue since 5.0.x series then this issue and the [#13806](https://github.com/mautic/mautic/issues/13806) are duplicates. Therefore,  the solution you finally take would be the same for both issues. Because I propose in my Issue another solution but but you surely know Mautic better.

**Comment by pm-pmaas** [comment]
@avara so what needs be done to fix this issue ?

**Comment by abhisekmazumdar** [comment]
Yes I can confirm that moving the `$this->emailStatModel->saveEntity($stat)`  worked. Thanks everyone for the inputs

**Comment by RCheesley** [comment]
Hi folks, please test #13915 🚀

**Comment by pm-pmaas** [comment]
Is it true this does not correct the previously send emails ?

**Comment by escopecz** [comment]
That is true, but it could be fixed by a migration or even one time update query. At least for the unique counts.

**Comment by avara** [comment]
> @avara so what needs be done to fix this issue ?

Sorry for not replying to you before @pm-pmaas. But I'm glad to know that now you have all the answers you need.

**Comment by pm-pmaas** [comment]
> That is true, but it could be fixed by a migration or even one time update query. At least for the unique counts.

Somebody here who can make a fix to also correct the previously send e-mails ? I can confirm for new e-mails everything is working as expected again.

![Scherm­afbeelding 2024-07-04 om 11 01 56](https://github.com/mautic/mautic/assets/17729447/2d5e3e18-5cc6-44df-a0e1-a4ce3d80daa0)

**Comment by micaelomota** [comment]
Can't wait for this fix to be released 🔥

**Comment by escopecz** [comment]
The fix https://github.com/mautic/mautic/pull/13915 was merged and will be released in Mautic 5.1.1. I'm closing this issue now.
